File: dongle/frecency.py
"""
Frecency tracking: combines visit frequency and recency to rank directories.
Inspired by Mozilla's frecency algorithm and the `z` shell tool.
"""
import json
import time
from pathlib import Path

from dongle.config import FRECENCY_FILE

_MAX_ENTRIES = 1000


def _load() -> dict:
    if not FRECENCY_FILE.exists():
        return {}
    try:
        return json.loads(FRECENCY_FILE.read_text())
    except Exception:
        return {}


def _save(data: dict):
    try:
        FRECENCY_FILE.write_text(json.dumps(data))
    except Exception:
        pass


def record_visit(path: str):
    """Increment visit count and update last-visited timestamp for path."""
    data = _load()
    now = time.time()
    entry = data.get(path)
    if entry:
        entry["visits"] += 1
        entry["last_visit"] = now
    else:
        data[path] = {"visits": 1, "last_visit": now}

    # Keep only the most recently visited entries
    if len(data) > _MAX_ENTRIES:
        trimmed = sorted(data.items(), key=lambda x: x[1]["last_visit"], reverse=True)
        data = dict(trimmed[:_MAX_ENTRIES])

    _save(data)


def get_frecency_scores() -> dict:
    """
    Return {path: score} for all tracked directories.

    Recency weights:
      < 1 hour  → ×4
      < 1 day   → ×2
      < 1 week  → ×1
      older     → ×0.5
    """
    data = _load()
    if not data:
        return {}

    now = time.time()
    scores = {}
    for path, info in data.items():
        visits = info.get("visits", 1)
        age = now - info.get("last_visit", 0)

        if age < 3_600:
            weight = 4
        elif age < 86_400:
            weight = 2
        elif age < 604_800:
            weight = 1
        else:
            weight = 0.5

        scores[path] = int(visits * weight)

    return scores


def get_recent_dirs(n: int = 10) -> list:
    """Return up to n most recently visited directories."""
    data = _load()
    if not data:
        return []
    entries = sorted(data.items(), key=lambda x: x[1]["last_visit"], reverse=True)
    return [path for path, _ in entries[:n]]

File: lib/src/frecency.dart
import 'dart:convert';
import 'config.dart';

class Frecency {
  static const int _maxEntries = 1000;
  final DongleConfig config = DongleConfig();

  Map<String, dynamic> _load() {
    if (!config.frecencyFile.existsSync()) {
      return {};
    }
    try {
      final content = config.frecencyFile.readAsStringSync();
      return jsonDecode(content) as Map<String, dynamic>;
    } catch (e) {
      return {};
    }
  }

  void _save(Map<String, dynamic> data) {
    try {
      config.frecencyFile.writeAsStringSync(jsonEncode(data));
    } catch (e) {
      // ignore write errors
    }
  }

  void recordVisit(String path) {
    var data = _load();
    final now = DateTime.now().millisecondsSinceEpoch / 1000.0;

    if (data.containsKey(path)) {
      data[path]['visits'] = (data[path]['visits'] as int) + 1;
      data[path]['last_visit'] = now;
    } else {
      data[path] = {'visits': 1, 'last_visit': now};
    }

    if (data.length > _maxEntries) {
      var entries = data.entries.toList()
        ..sort(
          (a, b) => (b.value['last_visit'] as double).compareTo(
            a.value['last_visit'] as double,
          ),
        );

      var trimmed = <String, dynamic>{};
      for (var i = 0; i < _maxEntries; i++) {
        trimmed[entries[i].key] = entries[i].value;
      }
      data = trimmed;
    }

    _save(data);
  }

  Map<String, int> getFrecencyScores() {
    final data = _load();
    if (data.isEmpty) return {};

    final now = DateTime.now().millisecondsSinceEpoch / 1000.0;
    Map<String, int> scores = {};

    data.forEach((path, info) {
      final visits = (info['visits'] as num?)?.toInt() ?? 1;
      final lastVisit = (info['last_visit'] as num?)?.toDouble() ?? 0.0;
      final age = now - lastVisit;

      double weight;
      if (age < 3600) {
        weight = 4;
      } else if (age < 86400) {
        weight = 2;
      } else if (age < 604800) {
        weight = 1;
      } else {
        weight = 0.5;
      }

      scores[path] = (visits * weight).toInt();
    });

    return scores;
  }

  List<String> getRecentDirs([int n = 10]) {
    final data = _load();
    if (data.isEmpty) return [];

    var entries = data.entries.toList()
      ..sort(
        (a, b) => (b.value['last_visit'] as num).compareTo(
          a.value['last_visit'] as num,
        ),
      );

    return entries.take(n).map((e) => e.key).toList();
  }
}
